src/Umbraco.Core/Models/PublishedContent/PublishedContentTypeFactory.cs

```diff
@@ -73,24 +66,31 @@ public PublishedDataType GetDataType(int id)
     }
 
     /// <inheritdoc />
-    public void NotifyDataTypeChanges(int[] ids)
+    public void NotifyDataTypeChanges(params int[] ids)
     {
+        if (_publishedDataTypes is null)
+        {
+            // Not initialized yet, so skip and avoid lock
+            return;
+        }
+
         lock (_publishedDataTypesLocker)
         {
-            if (_publishedDataTypes == null)
+            if (ids.Length == 0)
             {
-                IEnumerable<IDataType> dataTypes = _dataTypeService.GetAll();
-                _publishedDataTypes = dataTypes.ToDictionary(x => x.Id, CreatePublishedDataType);
+                // Clear cache (and let it lazy initialize again later)
+                _publishedDataTypes = null;
             }
             else
             {
+                // Remove items from cache (in case the data type is removed)
                 foreach (var id in ids)
                 {
                     _publishedDataTypes.Remove(id);
                 }
 
-                IEnumerable<IDataType> dataTypes = _dataTypeService.GetAll(ids);
-                foreach (IDataType dataType in dataTypes)
+                // Update cacheB
```

Comment:
Let's fix this typo if other changes are required:
```suggestion
                // Update cache
```